Lib/test/test_os.py

```diff
@@ -2873,6 +2873,51 @@ def test_getfinalpathname_handles(self):
 
         self.assertEqual(0, handle_delta)
 
+    @support.requires_subprocess()
+    def test_stat_unlink_race(self):
+        # bpo-46785: the implementation of os.stat() falls back to reading
+        # the parent directory if CreateFileW() fails with a permission
+        # error. If reading the parent directory fails because the file or
+        # directory are subsequently unlinked, or because the volume or
+        # share are no longer available, then the original permission error
+        # should not be restored.
+        filename =  os_helper.TESTFN
+        self.addCleanup(os_helper.unlink, filename)
+        command = '''if 1:
+            import os
+            import sys
+            filename = sys.argv[1]
+            while True:
+                try:
+                    with open(filename, "w") as f:
+                        pass
+                except OSError:
+                    pass
+                try:
+                    os.remove(filename)
+                except OSError:
+                    pass
+        '''
+        ignored_errors = (
+            2,  # ERROR_FILE_NOT_FOUND
+            3,  # ERROR_PATH_NOT_FOUND
+            21, # ERROR_NOT_READY
+            67, # ERROR_BAD_NET_NAME
+        )
+        deadline = time.time() + 5
+        p = subprocess.Popen([sys.executable, '-c', command, filename])
+
+        try:
+            while time.time() < deadline:
+                try:
+                    os.stat(filename)
+                except OSError as e:
+                    if e.winerror not in ignored_errors:
```

Comment:
What do you think about generalizing this to ignore `FileNotFoundError`, which is raised for 3 of the 4 errors. The only case that's tested is `ERROR_FILE_NOT_FOUND`. 

`ERROR_NOT_READY` is handled as a `PermissionError`, which is nonsense given the reason for the error is almost always a removed disk/card from a volume that has removable media. But Python's "PC/errmap.h" follows what the C runtime does, which maps `ERROR_NOT_READY` to `EACCES`.

Reply:
Makes sense to me, since as you wrote, this only tests what happens when `ERROR_FILE_NOT_FOUND` is raised.

By the way, I only read some of the tests of `os.stat()`, but it seems like more edge cases relating to permission errors should be written. In the context of this PR, maybe we could also write a test which triggers `ERROR_PATH_NOT_FOUND` by removing the directory containing the "foo" file, while running `os.stat()` on the file?

Reply:
> In the context of this PR, maybe we could also write a test which triggers `ERROR_PATH_NOT_FOUND` by removing the directory containing the "foo" file, while running `os.stat()` on the file?

I implemented this (`test_stat_rmdir_race`), let me know what you think.